OpenChange/MAPIStoreMailMessage.m

```diff
@@ -396,36 +401,82 @@ - (void) _fetchBodyData
             encoding = @"7-bit";
 
           /* We should provide a case for each of the types in acceptedMimeTypes */
-          if ([mimeType isEqualToString: @"text/html"] && !mailIsEvent)
-            {
-              content = [content bodyDataFromEncoding: encoding];
-              [htmlContent appendData: content];
-            }
-          else if ([mimeType isEqualToString: @"text/plain"] && !mailIsEvent)
+          if (!mailIsEvent)
             {
+              NSString *charset;
+              BOOL html;
+              if ([mimeType isEqualToString: @"text/html"])
+                {
+                  html = YES;
+                }
+              else if ([mimeType isEqualToString: @"text/plain"])
+                {
+                  html = NO;
+                }
+              else
+                {
+                  [self warnWithFormat: @"Unsupported MIME type for non-event body part: %@.",
+                        mimeType];
+                  continue;
+                }
+                          
               content = [content bodyDataFromEncoding: encoding];
-              NSString *charset = [bodyPartsCharsets objectForKey: key];
+              charset = [bodyPartsCharsets objectForKey: key];
               if (charset)
                 {
-                  NSString *stringValue = [content bodyStringFromCharset: charset];
-                  [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                  NSString *stringValue = nil;
+                  if (html)
+                    {
+                      if ([charset isEqualToString: headerCharset])
+                        {
+                          [htmlContent appendData: content];
+                        }
+                      else
+                        {
+                          stringValue = [content bodyStringFromCharset: charset];
+                          [htmlContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                        }
+                      if (multipartMixed)
+                        {
+                          if (stringValue == nil)
+                            stringValue = [content bodyStringFromCharset: charset];
+                          [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                        }
+                    }
+                  else
+                    {
+                      /* No HTML, then is text/plain */
+                      stringValue = [content bodyStringFromCharset: charset];
+                      [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                      if (multipartMixed)
+                        {
+                          stringValue = [stringValue stringByReplacingOccurrencesOfString: @"\n"
```

Comment:
Whether return HTML or text is the client to choose. We fill the two NSMutableData, one for the html and other for the text.
In case of multipart we check that if it is mixed or alternative. If it is alternative it means that there is a different representation of the same message, so we put html parts in html data and text parts in text data.
If it is mixed we must assemble a message from its parts, so we put in the two. In text case we are transforming eol in br marks to increase readability.